manager/controllers/app/plotter_controller.go

```diff
@@ -120,17 +124,241 @@ func (r *PlotterReconciler) reconcileFinalizers(plotter *app.Plotter) error {
+					stepTemplate := plotter.Spec.Templates[seqStep.Template]
+					// isPrimaryModule := true
+					for _, module := range stepTemplate.Modules {
+						moduleArgs := seqStep.Parameters
+
+						// If the module type is "plugin" then it is assumed
+						// that there is a primary module of type "config" or "service"
+						// in the same template and all the module arguments are used only by
+						// the primary module
+						if module.Type == "plugin" {
+							moduleArgs = nil
+						}
+						scope := r.getModuleScope(module.Capabilities, subFlow.FlowType)
+						clusterName := seqStep.Cluster
+						var authPath string
+						for _, cluster := range clusters {
+							if clusterName == cluster.Name {
+								authPath = utils.GetAuthPath(cluster.Metadata.VaultAuthPath)
+								break
+							}
+						}
+						plotterModule := PlotterModulesSpec{
+							ModuleArguments: moduleArgs,
+							AssetID:         flow.AssetID,
+							FlowType:        subFlow.FlowType,
+							ClusterName:     clusterName,
+							Chart:           module.Chart,
+							ModuleName:      module.Name,
+							Scope:           scope,
+							VaultAuthPath:   authPath,
+						}
+
+						blueprintModule := r.convertPlotterModuleToBlueprintModule(plotter, plotterModule)
+						// append the module to the modules list
+						moduleInstances = append(moduleInstances, *blueprintModule)
+					}
+				}
+			}
+		}
+	}
+	blueprints := r.GenerateBlueprints(moduleInstances)
+
+	return blueprints
+}
+
+// updatePlotterAssetsState updates the status of the assets processed by the blueprint modules.
+func (r *PlotterReconciler) updatePlotterAssetsState(assetToStatusMap map[string]app.ObservedState, blueprint *app.Blueprint) {
+	for instanceName, moduleState := range blueprint.Status.ModulesState {
+		for _, assetID := range blueprint.Spec.Modules[instanceName].AssetIDs {
+			state, exists := assetToStatusMap[assetID]
+			errMsg := moduleState.Error
+			if exists {
+				errMsg = state.Error + "\n" + errMsg
+			}
+			if !exists {
+				assetToStatusMap[assetID] = moduleState
+				// if the current module is not ready then update all its assets
+				// to not ready state regardless of the assets state
+			} else if !moduleState.Ready {
+				assetToStatusMap[assetID] = app.ObservedState{
+					Ready: false,
+					Error: errMsg,
+				}
+			}
+		}
+	}
+}
+
+// setPlotterAssetsReadyStateToFalse sets to false the status of the assets processed by the blueprint modules.
+func (r *PlotterReconciler) setPlotterAssetsReadyStateToFalse(assetToStatusMap map[string]app.ObservedState, blueprintSpec *app.BlueprintSpec, errMsg string) {
+	for _, module := range blueprintSpec.Modules {
+		for _, assetID := range module.AssetIDs {
+			var err = errMsg
+			state, exists := assetToStatusMap[assetID]
+			if exists {
+				err = state.Error + err
+				delete(assetToStatusMap, assetID)
+			}
+
+			assetToStatusMap[assetID] = app.ObservedState{
+				Ready: false,
+				Error: err,
+			}
+		}
```

Comment:
Just this assignment is fine. You don't need to check for existence and delete the old value from the map.

Reply:
Fixed